aws/sns.go

```diff
@@ -0,0 +1,102 @@
+package aws
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/sns"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/hashicorp/go-multierror"
+)
+
+func getAllSNSTopics(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	cfg, err := awsconfig.LoadDefaultConfig(context.TODO(), awsconfig.WithRegion(aws.StringValue(session.Config.Region)))
+	if err != nil {
+		return []*string{}, errors.WithStackTrace(err)
+	}
+	svc := sns.NewFromConfig(cfg)
+
+	allSNSTopics := []*string{}
+
+	paginator := sns.NewListTopicsPaginator(svc, nil)
+
+	for paginator.HasMorePages() {
+		resp, err := paginator.NextPage(context.TODO())
+		if err != nil {
+			return []*string{}, errors.WithStackTrace(err)
+		}
+		for _, topic := range resp.Topics {
+			allSNSTopics = append(allSNSTopics, topic.TopicArn)
+		}
+	}
+	return allSNSTopics, nil
+}
+
+func nukeAllSNSTopics(session *session.Session, identifiers []*string) error {
+	region := aws.StringValue(session.Config.Region)
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.TODO(), awsconfig.WithRegion(aws.StringValue(session.Config.Region)))
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	svc := sns.NewFromConfig(cfg)
+
+	if len(identifiers) == 0 {
+		logging.Logger.Infof("No SNS Topics to nuke in region %s", region)
+	}
+
+	if len(identifiers) > 100 {
+		logging.Logger.Errorf("Nuking too many SNS Topics (100): halting to avoid hitting AWS API rate limiting")
+		return TooManySNSTopicsErr{}
+	}
+
```

Comment:
Do we know that 100 is the limit? Should we detect rate limiting and respond to that, or is that overkill?

Reply:
We don't know for sure, but this is the pattern for proactively trying to avoid rate limiting that I find us carrying throughout the resources. It would be nice if we had some intelligent logic to respond to rate limiting, but I'd need to think about the cleanest way to do that.